In [2]:
pip install pyECLAT

Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'C:\Users\91976\anaconda3\python.exe -m pip install --upgrade pip' command.


In [3]:
from pyECLAT import Example1, Example2
ex2 = Example2().get()
ex2

0                  1                  2  \
0                shrimp            almonds            avocado   
1               burgers          meatballs               eggs   
2               chutney                NaN                NaN   
3                turkey            avocado                NaN   
4         mineral water               milk         energy bar   
...                 ...                ...                ...   
2996          green tea                NaN                NaN   
2997             shrimp               cake                NaN   
2998                ham        ground beef      mineral water   
2999            burgers  frozen vegetables  whole wheat pasta   
3000  whole wheat pasta               eggs               cake   

                     3             4                 5            6  
0       vegetables mix  green grapes  whole weat flour         yams  
1                  NaN           NaN               NaN          NaN  
2                  NaN           NaN               NaN          NaN  
3                  NaN           NaN               NaN          NaN  
4     whole wheat rice     green tea               NaN          NaN  
...                ...           ...               ...          ...  
2996               NaN           NaN               NaN          NaN  
2997               NaN           NaN               NaN          NaN  
2998               NaN           NaN               NaN          NaN  
2999     mineral water     chocolate              eggs  blueberries  
3000      french fries     green tea               NaN          NaN  

[3001 rows x 7 columns]

In [4]:
from pyECLAT import ECLAT
eclat_instance = ECLAT(ex2, verbose=True)

100%|██████████████████████████████████████████████████████████████████████████████| 119/119 [00:00<00:00, 2483.87it/s]


In [5]:
get_ECLAT_indexes, get_ECLAT_supports = eclat_instance.fit(min_support=0.030,
                                                           min_combination=1,
                                                           max_combination=3,
                                                           separator=' & ',
                                                           verbose=True)

print(get_ECLAT_supports)

1it [00:00,  5.79it/s]

Combination 1 by 1


32it [00:00, 58.83it/s]
17it [00:00, 163.15it/s]

Combination 2 by 2


496it [00:03, 151.61it/s]
28it [00:00, 140.83it/s]

Combination 3 by 3


4960it [00:37, 132.57it/s]

{'escalope': 0.07064311896034656, 'whole wheat rice': 0.045318227257580806, 'low fat yogurt': 0.05664778407197601, 'cooking oil': 0.04798400533155615, 'shrimp': 0.0773075641452849, 'avocado': 0.035654781739420195, 'ground beef': 0.09363545484838387, 'eggs': 0.17727424191936023, 'honey': 0.03865378207264245, 'frozen vegetables': 0.09196934355214928, 'whole wheat pasta': 0.03032322559146951, 'fresh bread': 0.03032322559146951, 'green tea': 0.11329556814395202, 'turkey': 0.06597800733088971, 'mineral water': 0.23692102632455847, 'spaghetti': 0.18293902032655782, 'burgers': 0.08563812062645784, 'frozen smoothie': 0.04798400533155615, 'cookies': 0.07697434188603798, 'milk': 0.12695768077307565, 'champagne': 0.042652449183605466, 'salmon': 0.03898700433188937, 'pancakes': 0.07864045318227257, 'grated cheese': 0.055314895034988334, 'tomatoes': 0.06864378540486504, 'herb & pepper': 0.05031656114628457, 'chocolate': 0.1616127957347551, 'cake': 0.07697434188603798, 'french fries': 0.154281906031

# Itemsets:
    'spaghetti & frozen vegetables': 0.030989670109963344, 
    'spaghetti & mineral water': 0.06064645118293902, 
    'spaghetti & ground beef': 0.03598800399866711, 
    'spaghetti & chocolate': 0.03732089303565478, 
    'spaghetti & eggs': 0.03265578140619793, 
    'spaghetti & milk': 0.03865378207264245, 
    'frozen vegetables & mineral water': 0.03698767077640786, 
    'french fries & eggs': 0.03432189270243252, 
    'mineral water & ground beef': 0.03698767077640786, 
    'mineral water & chocolate': 0.04765078307230923, 
    'mineral water & eggs': 0.04798400533155615, 
    'mineral water & milk': 0.048317227590803064
    


In [7]:
import pandas as pd
get_ECLAT_indexes, get_ECLAT_supports = eclat_instance.fit(min_support=0.036,
                                                           min_combination=2,
                                                           max_combination=2,
                                                           separator=' & ',
                                                           verbose=True)
print(get_ECLAT_supports)
rank = pd.DataFrame(get_ECLAT_supports,index=[0])
rank.transpose().head(1)

13it [00:00, 128.84it/s]

Combination 2 by 2


406it [00:02, 144.43it/s]

{'ground beef & mineral water': 0.03698767077640786, 'eggs & mineral water': 0.04798400533155615, 'frozen vegetables & mineral water': 0.03698767077640786, 'mineral water & spaghetti': 0.06064645118293902, 'mineral water & milk': 0.048317227590803064, 'mineral water & chocolate': 0.04765078307230923, 'spaghetti & milk': 0.03865378207264245, 'spaghetti & chocolate': 0.03732089303565478}


0
ground beef & mineral water  0.036988

# Itemsets:
    'spaghetti & mineral water': 0.06064645118293902, 
    'spaghetti & ground beef': 0.03598800399866711, 
    'spaghetti & chocolate': 0.03732089303565478, 
    'spaghetti & milk': 0.03865378207264245, 
    'frozen vegetables & mineral water': 0.03698767077640786, 
    'french fries & eggs': 0.03432189270243252, 
    'mineral water & ground beef': 0.03698767077640786, 
    'mineral water & chocolate': 0.04765078307230923, 
    'mineral water & eggs': 0.04798400533155615, 
    'mineral water & milk': 0.048317227590803064

In [8]:
get_ECLAT_indexes, get_ECLAT_supports = eclat_instance.fit(min_support=0.037,
                                                           min_combination=1,
                                                           max_combination=4,
                                                           separator=' & ',
                                                           verbose=True)

print(get_ECLAT_supports)

15it [00:00, 71.57it/s]

Combination 1 by 1


29it [00:00, 75.31it/s]
14it [00:00, 131.92it/s]

Combination 2 by 2


406it [00:02, 140.50it/s]
11it [00:00, 103.08it/s]

Combination 3 by 3


3654it [00:27, 134.95it/s]
9it [00:00, 85.13it/s]

Combination 4 by 4


23751it [03:30, 112.77it/s]

{'escalope': 0.07064311896034656, 'whole wheat rice': 0.045318227257580806, 'low fat yogurt': 0.05664778407197601, 'cooking oil': 0.04798400533155615, 'shrimp': 0.0773075641452849, 'ground beef': 0.09363545484838387, 'eggs': 0.17727424191936023, 'honey': 0.03865378207264245, 'frozen vegetables': 0.09196934355214928, 'green tea': 0.11329556814395202, 'turkey': 0.06597800733088971, 'mineral water': 0.23692102632455847, 'spaghetti': 0.18293902032655782, 'burgers': 0.08563812062645784, 'frozen smoothie': 0.04798400533155615, 'cookies': 0.07697434188603798, 'milk': 0.12695768077307565, 'champagne': 0.042652449183605466, 'salmon': 0.03898700433188937, 'pancakes': 0.07864045318227257, 'grated cheese': 0.055314895034988334, 'tomatoes': 0.06864378540486504, 'herb & pepper': 0.05031656114628457, 'chocolate': 0.1616127957347551, 'cake': 0.07697434188603798, 'french fries': 0.15428190603132289, 'olive oil': 0.06597800733088971, 'soup': 0.054315228257247584, 'chicken': 0.05664778407197601, 'eggs & 

# Itemsets:
    'spaghetti & mineral water': 0.06064645118293902, 
    'spaghetti & chocolate': 0.03732089303565478, 
    'spaghetti & milk': 0.03865378207264245, 
    'mineral water & chocolate': 0.04765078307230923, 
    'mineral water & eggs': 0.04798400533155615, 
    'mineral water & milk': 0.048317227590803064

Algorithm to distinguish seasonal product
SeasonalDependency (topN, months)

for all pi ← topN

do

if si > = MPV

then

if MR (pi, month) < TR

then

RemovefromList (topN, pi)


In [9]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

In [10]:
# Data Loading
dataset = pd.read_csv('Market_Basket_Optimisation.csv', header = None)

# Adding all customers into a list of lists
transactions = []
for i in range(0, 7501):
    transactions.append([str(dataset.values[i,j]) for j in range(0, 20)])

In [11]:
import json
file = 'test.json'
with open(file) as rental_file:
    dict_test = json.load(rental_file)

# converting json dataset from dictionary to dataframe
rented = pd.DataFrame.from_dict(dict_test,orient="index")
rented.reset_index(level=0, inplace=True)
new_rented = rented.drop(['index'], axis = 1)
new_rented 

0             1                          2
0          milk frother      raincoat               milk frother
1                 crocs  french press               french press
2             generator      gumboots  water proof phone dockets
3         clothes dryer   windsheater                  umberalla
4    mosquito repellent         crocs  water proof phone dockets
..                  ...           ...                        ...
195  mosquito repellent         crocs                   gumboots
196            raincoat     umberalla                  generator
197        coffee maker      gumboots  water proof phone dockets
198             frother       frother                   gumboots
199           bag cover     generator                  umberalla

[200 rows x 3 columns]

In [12]:
from pyECLAT import ECLAT
eclat_instance = ECLAT(new_rented, verbose=True)
get_ECLAT_indexes, get_ECLAT_supports = eclat_instance.fit(min_support=0.042,
                                                           min_combination=2,
                                                           max_combination=4,
                                                           separator=' & ',
                                                           verbose=False)
rank = pd.DataFrame(get_ECLAT_supports, index=[0])
final = rank.head(1)
arr = final.columns.values.tolist()
arr

100%|████████████████████████████████████████████████████████████████████████████████| 15/15 [00:00<00:00, 1664.45it/s]


['water proof phone dockets & umberalla',
 'water proof phone dockets & heater',
 'umberalla & clothes dryer',
 'mosquito repellent & crocs',
 'mosquito repellent & raincoat',
 'coffee maker & clothes dryer',
 'bag cover & clothes dryer']